### PR TITLE
tinyMCE use document "scriptoptions", and allow to override tinyMCE Javascript parameters

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue #10996 .
### Summary of Changes

The patch make tinyMCE plugin use `JDocument::addScriptOptions`/`JDocument::getScriptOptions` that automatically allow to override the editor options before the header rendering.

Another improve.
The patch also allow to have multiple editors on the page with specified options for each.
#### step 1

Check whether editor works as before. There was a lot changes in code, I could miss something.
#### step 2 Test override

In any system plugin for event `onBeforeRender`, try next code:

``` php
public function onBeforeRender()
{
    $doc = JFactory::getDocument();
    $editorOptions = $doc->getScriptOptions('plg_editor_tinymce');

    if(empty($editorOptions['tinyMCE']))
    {
        return;
    }

    $editorOptions['tinyMCE']['default']['menubar'] = false;

    $doc->addScriptOptions('plg_editor_tinymce', $editorOptions);
}
```

**Expected result:**

The menu bar should be disabled.
#### step 2 Test multiple editors with separated options

Use "Custom" module, and add there 2 more editor field:

``` xml
<field name="test1" type="editor" label="test1"/>
<field name="test2" type="editor" label="test2"/>
```

Then onBeforeRender event do:

``` php
public function onBeforeRender()
{
    $doc = JFactory::getDocument();
    $editorOptions = $doc->getScriptOptions('plg_editor_tinymce');

    if(empty($editorOptions['tinyMCE']))
    {
        return;
    }

    // Specify options for 'test1' editor
    $editorOptions['tinyMCE']['test1']['menubar'] = false;
    $editorOptions['tinyMCE']['test1']['toolbar1']  = 'bold italic underline strikethrough';

    // For 'test2' use default options, but disable "menubar"
    $editorOptions['tinyMCE']['test2'] = $editorOptions['tinyMCE']['default'];
    $editorOptions['tinyMCE']['test2']['menubar'] = false;

    $doc->addScriptOptions('plg_editor_tinymce', $editorOptions);
}
```

**Expected result:**
- Default editor of the Custom module should stay as configured in the plugin.
- editor with label "test1" should be without "menubar" and with `bold italic underline` buttons
- editor with label "test2" should be without "menubar" and "toolbar1" should looks as in default editor

### CODE REVIEW

File: plugins/editors/tinymce/tinymce.php [hunk 1]
```diff
@@ -804,187 +806,131 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
...
 		 **/
 		if (!$this->app->client->mobile)
 		{
-			$smallButtons = 'toolbar_items_size: "small",';
+			$scriptOptions['toolbar_items_size'] = 'small';
 		}
 		elseif ($mobileVersion == false)
 		{
-			$smallButtons = '';
+			//$smallButtons = '';
```
Comment (zero-24) [inline plugins/editors/tinymce/tinymce.php]: So this `elseif` can be removed?

File: plugins/editors/tinymce/tinymce.php [hunk 2]
```diff
@@ -804,187 +806,131 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
...
 		}
 		else
 		{
-			$smallButtons = '';
+			//$smallButtons = '';
```
Comment (zero-24) [inline plugins/editors/tinymce/tinymce.php]: and this line too?

File: plugins/editors/tinymce/tinymce.php (hunk 1 above)
Comment (Fedik) [inline plugins/editors/tinymce/tinymce.php]: ah, yes, good question, I forgot this :smile: 
this piece of code looks very strange in original version also, need to check more, what this is doing

Comment (brianteeman) [inline plugins/editors/tinymce/tinymce.php]: Iirc it changes the size of the toolbar buttons when it's a mobile view so
that they are big enough for fat fingers

Comment (Fedik) [inline plugins/editors/tinymce/tinymce.php]: @brianteeman thanks for explanation

File: media/editors/tinymce/js/tinymce-init.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+		tinyMCE.init(options);
+	}
+
+	// Init on doomready
+	$(document).ready(function(){
+		Joomla.setupEditorsTinyMCE();
+
+    	// Init in subform field
+    	$(document).on('subform-row-add', function(event, row){
```
Comment (dgt41) [inline media/editors/tinymce/js/tinymce-init.js]: 👍 💯

**Comment by schnuti** [comment]
I have tested this item :white_check_mark: successfully on 7b4b16ca6cfde53aee4011fa9766a4f7b00daa82<br /><br />1. Tested article edit and custom module with all three tinyMCE settings (Simple, Advanced and Extended) All OK!
2. Activated a plugin that overrides the settings and tested again as 1. Override OK!
3. Changed the plugin to use three different settings as described. Tested with "extra fields" added as parameters to the Custom Module. I got one "normal" editor and two modified. -> OK!

I'll have some use of 3. as well. :+1:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by schnuti** [comment]
I'll help with some documentation if you give me a hint to where it belongs.

**Comment by schnuti** [comment]
Maybe the key of the array should be named something else than "default". There are now 3 different default settings involved. The plugin default ( = Advanced ), the built in tinyMCE default settings and now in the document. Replace with something like "jmce" ?

**Comment by Fedik** [comment]
`default` here in Joomla context, and contain info how the plugin is configured, here are parameters selected by user in the plugin  configuration.
And we do not work with default options of the tinyMCE.js script directly, in any way.
`jmce` are less obvious name :wink:

**Comment by schnuti** [comment]
@Fedik 
Well I got a little confused. Not important but I need some explanation.

Is this "default" used for your example field test1 or the tinyMCE defaults? For field2 you explicitely load this "default"

In the custom plugin:
`
$editorOptions['tinyMCE']['test2'] = $editorOptions['tinyMCE']['default'];
`

Result in the HTML:
`
"test1":{"menubar":false,"toolbar1":"bold italic underline strikethrough"}
,"test2":{"suffix":".min","baseURL ....
`

**Comment by Fedik** [comment]
well, maybe then better I will rename it to `global` :smile: 

`$editorOptions['tinyMCE']['default']` there all options chosen by user in the plugin configuration. So they used by *default* for all "tinyMCE" editors in Joomla. 

When you do:
```php
$editorOptions['tinyMCE']['default']['menubar'] = false;
```
You override option chosen by the user in the plugin configuration. And it will affect all "tinyMCE" editors in Joomla, except those you override by next example.

When you do:
```php
$editorOptions['tinyMCE']['test1'] = array();
$editorOptions['tinyMCE']['test1']['menubar'] = false;
$editorOptions['tinyMCE']['test1']['toolbar1'] = 'bold italic underline strikethrough';
```
You fully ignore options chosen by user in the plugin configuration. 

When you do:
```php
$editorOptions['tinyMCE']['test2'] = $editorOptions['tinyMCE']['default'];
$editorOptions['tinyMCE']['test2']['menubar'] = false;
```
You copy options chosen by user to an editor with name `test2`, and modify them (`['menubar'] = false`) only for this editor instance.

**Comment by schnuti** [comment]
@Fedik 
Thanks! That's what I was assuming. God  to get confirmed.

A simple test without copied options. (test1)
I changed tinyMCE to use <br /> instead of <p>. Test1 still uses the tinyMCE fallback (default) <p>. As expected. 
If you want ta add a button like charmap, you have to add a "plugin-string" with "charmap" as well. As expected.

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on bc785a33db939b523dbf87bf948f41f111da1562<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by dgt41** [comment]
@Fedik  you need to resync this as @wilsonge did merge another PR for some button movement

**Comment by Fedik** [comment]
I will check

**Comment by schnuti** [comment]
@Fedik 
Test faild  :confused:
I tested with current-staging and the test data.

1. Only one instance of the editor appears in custom module. Added 2 as in former test.
2. Unsetting the menubar gives a breaking error.  Edit the toolbar works.
`
editorOptions['tinyMCE']['default']['menubar'] = false;

Fatal error: Cannot use temporary expression in write context in D:\xampp\htdocs\joomla-cms-staging\plugins\system\mymce\mymce.php
` 
Did you change to use call-backs?

**Comment by schnuti** [comment]
@Fedik 
Sorry, too early in the morning. I made the change to the backend module and checked the frontend one!
All three instances appear with different toolbars and the error was a typo of mine!
All is well.

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on 7804f39e95d057a16291654cbc45dc1c95ca84ea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by dgt41** [comment]
@schnuti can you mark your test in the issue tracker: https://issues.joomla.org/tracker/joomla-cms/11157
So we can move this to RTC?

**Comment by schnuti** [comment]
I have tested this item :white_check_mark: successfully on 7804f39e95d057a16291654cbc45dc1c95ca84ea<br /><br />All is well since my last test. New test, same result.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by zero-24** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by dgt41** [comment]
@zero-24 can you put a 3.7 milestone here?

**Comment by zero-24** [comment]
Currently we need to ask @brianteeman or anyone from the PLT to do this

**Comment by zero-24** [comment]
We should also add the new feature label maybe

**Comment by schnuti** [comment]
@Fedik , @dgt41 and anyone else with ideas around those overrides. If you have some time, please read the question I added on the Google dev. list
[Testing override of TinyMCE settings](https://groups.google.com/forum/#!topic/joomla-dev-general/SS4K58McMSE)

**Comment by schnuti** [comment]
@Fedik 
I'm trying some more overrides and have found a problem. I want to load plugins from external source ( = not the tinymce plugin path). I want to use this for some available plugins. 
This works when I add it directly to a tinymce init but not when I add it in the override.
```
...],
external_plugins: { 'cbtinybuttons' : 'http://localhost/j3/media/com_mycom/js/cbtinybuttons.min.js'},
toolbar: '   ...
```
I've tried a lot with single and double quotes but tinymce breaks and no editor is loaded.

```
$external_plugins = '{ "cbtinybuttons" : "'.JUri::root(true).
    						'/media/com_mycom/js/cbtinybuttons.min.js"}';
$editorOptions['tinyMCE']['default']['external_plugins'] = $external_plugins;
```
Result in the HTML source:
```
..."}],"external_plugins":"{ \"cbtinybuttons\" : \"\/j3\/media\/com_oejcb\/js\/cbtinybuttons.min.js\"}","toolbar2...`
```
I see a difference that there are : included in the parameter. 
Any idea?

**Comment by Fedik** [comment]
@schnuti  In theory should be like:
```php
$editorOptions['tinyMCE']['default']['external_plugins']['myPlugin1'] = 'path/to/plugin.js';
$editorOptions['tinyMCE']['default']['external_plugins']['myPlugin2'] = 'path/to/another-plugin.js';
```

**Comment by schnuti** [comment]
@Fedik 
Thanks a lot!  :bowtie:
This works as well! I loaded two tinymce plugins from another path in the media folder.
Maybe my code above would also work but this is obviously the way to do it.
If someone else want to test it ->
I used the wrong JUri! It has to be without true i.e. JUri::root() and without the first slash /  . Write then the rest as Fedik says.

**Comment by schnuti** [comment]
Now I'll make another test. Not really a test, I will use it.
As I do not have any plugins that modifies  any of three textarea fields, I use TinyMce for inline editing of those fields in frontend.
I'll move that code to use this tinyMce override instead of the other special coding.
I'm only trying to prove, how useful it can be.   :yum:

**Comment by alimpam** [comment]
I have tested this item :white_check_mark: successfully on 7804f39e95d057a16291654cbc45dc1c95ca84ea<br /><br />- Tested int article edit in different modes and it works
- used changes for new PR (making Editor-Buttons installable, commit will follow )
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11157">issues.joomla.org/joomla-cms/11157</a>.</sub>

**Comment by Fedik** [comment]
synced, because of #8544

**Pull request merged**